docs/architecture/adr-016-validator-consensus-key-rotation.md

```diff
@@ -0,0 +1,109 @@
+# ADR 016: Validator Consensus Key Rotation
+
+## Changelog
+
+- 2019 Oct 23: Initial draft
+
+## Context
+
+Validator consensus key rotation feature has been discussed and requested for a long time, for the sake of safer validator 
+key management policy (e.g. https://github.com/tendermint/tendermint/issues/1136). So, we suggest one of the simplest form of
+validator consensus key rotation implementation mostly onto Cosmos-SDK.
+
+## Decision
+
+### Pseudo procedure for consensus key rotation
+
+- create new random consensus key.
+- create and broadcast a transaction with a `MsgRotateConsPubKey` that states the new consensus key is now coupled with the validator operator with signature from the validator's operator key.
+- old consensus key becomes unable to participate on consensus immediately after the update of key mapping state on-chain.
+- start validating with new consensus key.
+- validators using HSM and KMS should update the consensus key in HSM to use the new rotated key after the height `h` when `MsgRotateConsPubKey` committed to the blockchain.
```

Comment:
I opened a KMS issue about supporting more than one registered key per chain, and ways that Tendermint could signal which key to use (e.g. if it detects a misconfiguration or invalid signature when requesting a signature from the new key):

https://github.com/tendermint/kms/issues/368

I think it's important to be able to handle failures during key rotations without requiring manual intervention, although perhaps it isn't required for a key rotation MVP. Without a single place to coordinate flipping which key is active, the rollback procedure will be much more complex, IMO.

Reply:
1)

If i am understanding right, tendermint/kms#368 seems similar to something like an OTP, even though it will be not practically one time.

Trigger of automatic key rotation can be decided by latest block hash. We can encode such block hash into an integer, and then key rotation is decided by the remainder of the integer divided by some high integer number. Also we can design an algorithm so that the rotation happens statistically every N hours or N days.

But these ideas are too complex to put it in this ADR at the moment.

2) There will be definitely an only struct in SDK which stores the ChangeLog of consensus key rotation. But a rollback on this data seems very risky, because there exists high possibility that the validator has either lost or been stolen the key. So, intuitively, even though the state of chain being rolled back, the consensus key for rolled back chain on genesis should not be rolled back but should be provided by each validator manually with their signature by the operator key.

I cannot think of any simple automatic solution to this yet.

Reply:
That seems overly complex and it's not immediately clear to me how that would work. As far as I see it, one of the main issues to address is once the Tx is committed in a block, KMS will somehow have to be notified automatically (or manually?) that it should use the new key in the keyring.

Reply:
I think notifying KMS is strictly KMS related job on tendermint, and it is quite independent from this initial key rotation feature on "Cosmos-SDK". Our intention is to deploy the simplest form of rotation feature on "Cosmos-SDK" so that we make no change on tendermint.

I think, after implementation of this feature, KMS expert (like @tarcieri) can hack Cosmos-SDK to support such notification by monitoring the mapping struc between valoper and valconspub. It should monitor a struc in SDK because the mapping information is not stored in tendermint.

Reply:
Ideally I think Tendermint would include the public key to use for the signature in the `SignProposal` and `SignVote` messages, although it seems that may not work for an MVP.

Reply:
Yes. This ADR includes only the simplest form of consensus key rotation without considering multiple consensus key situation. We think it should be remain as a long term goal to pursue because it needs a breaking structural change on Tendermint about validator and consensus key abstract definition.

Reply:
adding below statement in the context section

_Also, it should be noted that this ADR includes only the simplest form of consensus key rotation without considering multiple consensus keys concept. Such multple consensus keys concept shall be remained as a long term goal of Tendermint and Cosmos-SDK._